src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -153,7 +165,8 @@ protected DynamicObject executeMatch(DynamicObject regexp, DynamicObject string,
             assert match >= 0;
 
             final Region region = matcher.getEagerRegion();
-            DynamicObject result = allocateNode.allocate(matchDataClass(), Layouts.MATCH_DATA.build(string,
+            final DynamicObject dupedString = (DynamicObject) dupNode.call(frame, string, "dup");
```

Comment:
You don't really need the frame here, you call `dup` ignoring visibility and passing `null` for the frame argument.
That would avoid the bigger diff.
(It's on my TODO list to have a nice PrivateCallNode or so to do this more nicely)